### ISSUE TITLE
Error when deleting a service identity

### ISSUE THREAD

**Issue by ScottGuymer** [philips-software/terraform-provider-hsdp#104]
When you have a scenario where you have to delete a service identity it fails with a 409 when that service id is allocated to groups. It seems it will not do a cascade delete of the ID and the groups need to be removed first.

```
╷
│ Error: DELETE https://idm-client-test.us-east.philips-healthsuite.com/authorize/identity/Service/xXXXXXXXX-4a22e20ac189: 409
│
│
╵
```

You can see this error clearly in the UI when you try to do the delete
![image](https://user-images.githubusercontent.com/956825/129360051-5e5a7797-a632-4f48-96e7-2e3bb42e370e.png)

If you manually remove the group allocations and then run the terraform again it works as expected.

**Comment by ScottGuymer** [comment]
It was maybe a few minor versions older. The message will indeed be more helpful.

I can see how this might happen seeing how the group membership is modelled in the API.. Is there a way to tell the provider to modify the groups before deleting the service? (presuming the groups are managed in the same module)

**Comment by loafoe** [comment]
Not really, since the current IAM model is not conducive to discovering to which groups a service belongs to i.e. you would have to introspect the service using its own keys or do a brute force membership search, both options are not appealing.

**Comment by ScottGuymer** [comment]
Does TF not give you the explicit dependency between a service and the other resources it is referenced in?

**Comment by loafoe** [comment]
In the case of `group.devices = [device.id]` , the state only keeps track of member IDs from the groups perspective. 

The device does not know how its ID is used and there is are no "relational tools" in Terraform that will help you discover possible ID references. So, while technically the information might be there in some way, cascade delete in this case  would go against the declarative nature of TF i.e. how do you tell a group to remove a device without changing the HCL declaration of the group?

**Comment by ScottGuymer** [comment]
True.. 

but you would not be able to remove the declaration of a service without removing the references too it.

I think here the issue is in the order of operations.. 
It needs to modify (or delete) the group(s) before the service id.

not sure if it is fixable or not..

**Comment by loafoe** [comment]
If the device itself is added to a group which the Terraform module does not control, there is no way to delete the reference in that group. For one, you may not even have access to that group itself. 

This is kind of a fundamental problem in IAM i.e. if I know your service GUID I can potentially keep it hostage forever by assigning it to a group which is outside your sphere of influence. I will put this in front if the IAM SMEs for some feedback.

**Comment by ScottGuymer** [comment]
hmm yeah not great.

But we can at least try from within the provider.

If it removes/edits all of the group memberships first then at least it has a chance of working.

**Comment by loafoe** [comment]
This should resolve naturally inside a (root) module, as long as you are using the right references and e.g. not using literal string values, but using `attribute` references. 

There is also the limitation that AFAIK currently there is no API call to get a list of services associated to a group.

**Comment by ScottGuymer** [comment]
im not sure its resolving correctly

As I have been creating all the groups and users in the same tf module.

Will have to test again at some point and see if i can create an example.
